go/consensus/cometbft/apps/keymanager/churp/txs.go

```diff
@@ -304,6 +290,128 @@ func (ext *churpExt) apply(ctx *tmapi.Context, req *churp.SignedApplicationReque
 	return nil
 }
 
+func (ext *churpExt) confirm(ctx *tmapi.Context, req *churp.SignedConfirmationRequest) error {
+	// Prepare state.
+	state := churpState.NewMutableState(ctx.State())
+
+	// Ensure that the runtime exists and is a key manager.
+	kmRt, err := common.KeyManagerRuntime(ctx, req.Confirmation.RuntimeID)
+	if err != nil {
+		return err
+	}
+
+	// Get the existing status.
+	status, err := state.Status(ctx, req.Confirmation.RuntimeID, req.Confirmation.ID)
+	if err != nil {
+		return fmt.Errorf("keymanager: churp: non-existing ID: %d", req.Confirmation.ID)
+	}
+
+	// Allow confirmations only during the next handoff.
+	now, err := ext.state.GetCurrentEpoch(ctx)
+	if err != nil {
+		return err
+	}
+
+	switch status.NextHandoff {
+	case churp.HandoffsDisabled:
+		return fmt.Errorf("keymanager: churp: handoffs disabled")
+	case now:
+	default:
+		return fmt.Errorf("keymanager: churp: confirmations closed")
+	}
+
+	if status.NextHandoff != req.Confirmation.Epoch {
+		return fmt.Errorf("keymanager: churp: invalid handoff: got %d, expected %d", req.Confirmation.Epoch, status.NextHandoff)
+	}
+
+	// Check that application exists.
+	nodeID := ctx.TxSigner()
+	app, ok := status.Applications[nodeID]
+	if !ok {
+		return fmt.Errorf("keymanager: churp: application not found")
+	}
+
+	// Allow only one confirmation per handoff.
+	if app.Reconstructed {
+		return fmt.Errorf("keymanager: churp: confirmation already submitted")
+	}
+
+	// Verify checksum.
+	switch status.NextChecksum {
+	case nil:
+		// The first node to confirm is the source of truth.
+		status.NextChecksum = &req.Confirmation.Checksum
+	default:
+		// Other nodes need to confirm with the same checksum.
+		if !req.Confirmation.Checksum.Equal(status.NextChecksum) {
+			return fmt.Errorf("keymanager: churp: checksum mismatch: got %s, expected %s", req.Confirmation.Checksum, status.NextChecksum)
+		}
+	}
+
+	// Verify RAK signature.
+	rak, err := runtimeAttestationKey(ctx, nodeID, now, kmRt)
+	if err != nil {
+		return err
+	}
+	if err = req.VerifyRAK(rak); err != nil {
+		return fmt.Errorf("keymanager: churp: invalid signature: %w", err)
+	}
+
+	if ctx.IsCheckOnly() {
+		return nil
+	}
+
+	// Charge gas for this operation.
+	kmParams, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return err
+	}
+	if err = ctx.Gas().UseGas(1, churp.GasOpConfirm, kmParams.GasCosts); err != nil {
+		return err
+	}
+
+	// Return early if simulating since this is just estimating gas.
+	if ctx.IsSimulation() {
+		return nil
+	}
+
+	// Update application.
+	status.Applications[nodeID] = churp.Application{
+		Checksum:      app.Checksum,
+		Reconstructed: true,
+	}
+
+	// Try to finalize the handoff.
+	handoffCompleted := true
+	for _, app := range status.Applications {
+		if !app.Reconstructed {
+			handoffCompleted = false
+			break
+		}
+	}
+	if handoffCompleted {
+		status.Handoff = status.NextHandoff
+		status.Checksum = status.NextChecksum
+		status.Committee = maps.Keys(status.Applications)
```

Comment:
This is non-deterministic and will cause consensus failures. You need to sort the entries.